ops/model.py

```diff
@@ -1348,6 +1348,51 @@ class InvalidStatusError(ModelError):
     """Raised if trying to set an Application or Unit status to something invalid."""
 
 
+def _create_action_set_dict(dictionary, parent_key=False):
+    """Turn a nested dictionary into a flattened dictionary, using '.' as a key seperator.
+
+    This is used to allow nested dictionaries to be translated into the dotted format required by
+    the Juju `action-set` hook tool in order to set nested data on an action.
+
+    Additionally, this method performs some validation on keys to ensure they only use permitted
+    characters, which according to the action-set docs, are lowercase characters, digits, hyphens,
+    and periods.
+
+    Example::
+
+        test_dict = {'a': {'b': 1, 'c': 2}}
+        flat_dict = flatten(test_dict)
+        # flat_dict is now {'a.b': 1, 'a.c': 2}
+
+    Arguments:
+        dictionary: The dictionary to flatten
+        parent_key: The string to prepend to dictionary's keys
+
+    Returns:
+        A flattened dictionary
+
+    Raises:
+        ValueError: if a dict is passes with a key that fails to meet the format requirements
+    """
+    items = []
+    for key, value in dictionary.items():
+        # Ensure the key is of a valid format, and raise a ValueError if not
+        if not re.match(r'^[0-9a-z](([A-Za-z0-9-]+)?[a-z0-9])?$', key):
+            raise ValueError("Key '{}' is invalid: keys must start and end with lowercase "
+                             "alphanumeric, and contain only lowercase alphanumeric hyphens and "
+                             "periods".format(key))
```

Comment:
Periods too? I do not see a period in the regex.

Reply:
Per discussion elsewhere, if the intent is to not match on periods and to consider them reserved for the sake of sanely representing the layers of nesting (which I agree with), then this needs to be corrected to remove "periods".